components/raftstore/src/store/worker/pd.rs

```diff
@@ -1084,6 +1169,250 @@ fn send_destroy_peer_message(
+    pub split_score: f64,
+    pub interval: Duration,
+    pub min_sample_num: i32,
+}
+
+impl SplitHub {
+    pub fn new() -> SplitHub {
+        SplitHub {
+            region_qps: HashMap::default(),
+            region_keys: HashMap::default(),
+            region_recorder: HashMap::default(),
+            qps_threshold: DEFAULT_QPS_THRESHOLD,
+            split_score: DEFAULT_SPLIT_SCORE,
+            interval: DETECT_INTERVAL,
+            min_sample_num: MIN_SAMPLE_NUM,
+        }
+    }
+
+    fn add_qps(&mut self, region_id: u64, peer: &metapb::Peer, num: u32) {
+        let region_info = self
+            .region_qps
+            .entry(region_id)
+            .or_insert_with(RegionInfo::new);
+        region_info.add(peer, num);
+    }
+
+    fn add_key_range(&mut self, region_id: u64, start_key: &[u8], end_key: &[u8]) {
+        let key_ranges = self.region_keys.entry(region_id).or_insert_with(|| vec![]);
+        (*key_ranges).push(KeyRange::new(start_key, end_key));
+    }
+
+    pub fn add(&mut self, region_id: u64, peer: &metapb::Peer, start_key: &[u8], end_key: &[u8]) {
+        self.add_qps(region_id, peer, 1);
+        self.add_key_range(region_id, start_key, end_key);
+    }
+
+    fn update(&mut self, other: &mut SplitHub) {
+        for (region_id, region_info) in other.region_qps.iter() {
+            self.add_qps(*region_id, &(*region_info).peer, (*region_info).qps);
+        }
+        for (region_id, other_key_ranges) in other.region_keys.iter_mut() {
+            let key_ranges = self.region_keys.entry(*region_id).or_insert_with(|| vec![]);
+            (*key_ranges).append(other_key_ranges);
+        }
+    }
+
+    fn clear(&mut self) {
+        self.region_keys.clear();
+        self.region_qps.clear();
+        self.region_recorder.retain(|_, recorder| {
+            recorder.create_time.elapsed().unwrap() < DETECT_INTERVAL * DETECT_TIMES * 10
+        });
+    }
+
+    pub fn flush(&mut self) -> (Vec<u32>, Vec<SplitInfo>) {
+        let mut split_infos = Vec::default();
+        let mut top = BinaryHeap::with_capacity(TOP_N as usize);
+        for (region_id, region_info) in self.region_qps.iter() {
+            let qps = (*region_info).qps;
+            if qps > self.qps_threshold {
+                let recorder = self
+                    .region_recorder
+                    .entry(*region_id)
+                    .or_insert_with(Recorder::new);
+                recorder.record(self.region_keys.get(region_id).unwrap());
+                let key = recorder.split_key(self.split_score, self.min_sample_num);
+                if !key.is_empty() {
+                    let split_info = SplitInfo {
+                        region_id: *region_id,
+                        split_key: Key::from_raw(&key).into_encoded(),
+                        peer: (*region_info).peer.clone(),
+                    };
+                    split_infos.push(split_info);
+                    self.region_recorder.remove(region_id);
```

Comment:
I think this region should also be removed even if there's no suitable key, because we always need newest qps statistics.

Reply:
Good